firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigAutoFetch.java

```diff
@@ -194,35 +200,78 @@ public void run() {
   }
 
   @VisibleForTesting
-  public synchronized void fetchLatestConfig(int remainingAttempts, long targetVersion) {
-    int currentAttempts = remainingAttempts - 1;
+  public synchronized Task<Void> fetchLatestConfig(int remainingAttempts, long targetVersion) {
+    int remainingAttemptsAfterFetch = remainingAttempts - 1;
+    int currentAttemptNumber = MAXIMUM_FETCH_ATTEMPTS - remainingAttemptsAfterFetch;
 
-    // fetchAttemptNumber is calculated by subtracting current attempts from the max number of
-    // possible attempts.
     Task<ConfigFetchHandler.FetchResponse> fetchTask =
         configFetchHandler.fetchNowWithTypeAndAttemptNumber(
-            ConfigFetchHandler.FetchType.REALTIME, MAXIMUM_FETCH_ATTEMPTS - currentAttempts);
-    fetchTask.onSuccessTask(
-        (fetchResponse) -> {
-          long newTemplateVersion = 0;
-          if (fetchResponse.getFetchedConfigs() != null) {
-            newTemplateVersion = fetchResponse.getFetchedConfigs().getTemplateVersionNumber();
-          } else if (fetchResponse.getStatus()
-              == ConfigFetchHandler.FetchResponse.Status.BACKEND_HAS_NO_UPDATES) {
-            newTemplateVersion = targetVersion;
-          }
+            ConfigFetchHandler.FetchType.REALTIME, currentAttemptNumber);
+    Task<ConfigContainer> activatedConfigsTask = activatedCache.get();
 
-          if (newTemplateVersion >= targetVersion) {
-            executeAllListenerCallbacks();
-          } else {
-            Log.d(
-                TAG,
-                "Fetched template version is the same as SDK's current version."
-                    + " Retrying fetch.");
-            // Continue fetching until template version number if greater then current.
-            autoFetch(currentAttempts, targetVersion);
-          }
-          return Tasks.forResult(null);
-        });
+    return Tasks.whenAllComplete(fetchTask, activatedConfigsTask)
+        .continueWithTask(
+            scheduledExecutorService,
+            (listOfUnusedCompletedTasks) -> {
+              if (!fetchTask.isSuccessful()) {
+                return Tasks.forException(
+                    new FirebaseRemoteConfigClientException(
+                        "Failed to auto-fetch config update.", fetchTask.getException()));
+              }
+
+              if (!activatedConfigsTask.isSuccessful()) {
```

Comment:
I'd prefer to return an error here rather than catch it since we might obscure a problem and return incomplete results (ex. a deleted param wouldn't show up).

Reply:
Yeah returning an error makes sense here but I just wanted to bring up a potential edge case where the fetch is successful but getting the activated configs is not. In this case the new fetched config/template version will be saved on the device but the user's listener wouldn't be notified of it and so it would seem like the app never got the Config Update but it actually did. It would need to wait until the next event that incremented the template version before it gets another invalidation signal. Doesn't need to be resolved now, just wanted to bring up the possibility.

Reply:
Just wanted to bring up this possible scenario

Reply:
Yeah... that's a good call out. Since we'd still surface the error, it seems right to raise it rather than change the SDK behavior in this case. Devs can still decide if they want to manually fetch/activate when they get an error from realtime.